numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -653,6 +654,24 @@ static PyObject *
 /**end repeat**/
 
 static PyObject *
+voidtype_repr(PyObject *self){
+    PyVoidScalarObject *s = (PyVoidScalarObject*) self;
+    if (PyDataType_HASFIELDS(s->descr)) {
+        return gentype_repr(self);
+    }
+    return _Py_strhex(s->obval, s->descr->elsize);
```

Comment:
Could fall back on the "invoke something in `_internal.py` strategy again" - repr doesn't need to be incredibly fast